### ISSUE TITLE
[MonoAOTCompiler] GenerateAotModulesTable should use assembly names not file names

### ISSUE THREAD

**Issue by lambdageek** [dotnet/runtime#53249]
The [`System.Runtime.Loader.DefaultContext.Tests.csproj`](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Runtime.Loader/tests/DefaultContext/System.Runtime.Loader.DefaultContext.Tests.csproj#L16) references `System.Runtime.Loader.Noop.Assembly` but with a filename of `System.Runtime.Loader.Noop.Assembly_test.dll`

As a result, the AOT compiler creates:
```c
extern void *mono_aot_module_System_Runtime_Loader_Noop_Assembly_test_info;
...
{
	mono_aot_register_module (mono_aot_module_System_Runtime_Loader_Noop_Assembly_test_info);
}
```

which uses the wrong symbol `_test_info` instead of `_info`.

The AOT compiler task should use MetadataLoadContext to open the assemblies and to get the AssemblyName and generate the symbol name from that instead of using the filename.

**Comment by lambdageek** [comment]
To repro.

Build the runtime for some fullaot target - for example `./build.sh --os maccatalyst --arch arm64 -c Release`
then in `src/libraries/System.Runtime.Loader/tests/DefaultContext` do `../../../../../dotnet.sh build /p:RunAOTCompilation=true /p:MonoForceInterpreter=false /t:Test /p:TargetOS=MacCatalyst -c Release /bl`

which will fail with a non-zero exit code from `Ld`.

**Comment by lambdageek** [comment]
Fixed by #53659
